Cray environment for CUDAHOSTCXX is goofed

<!-- Explain, in a clear and concise way, the command you ran and the result you were trying to achieve.
Example: "I ran `spack find` to list all the installed packages and ..." -->

### Steps to reproduce the issue

On CoriGPU I am building lbann using GCC 8.2.0 and cmake is unable to resolve CUDAHOSTCXX because spack setups up a cray environment even though I am not using one.  It ends up with CUDAHOSTCXX=CC, which is wrong.


### Error Message

<!-- If Spack reported an error, provide the error message. If it did not report an error but the output appears incorrect, provide the incorrect output. If there was no error message and no output but the result is incorrect, describe how it does not match what you expect. -->

```console
CMake Error at /global/u2/v/vanessen/spack.git/opt/spack/cray-cnl7-skylake_avx512/gcc-8.2.0/cmake-3.18.0-esuryqzyhasr24bju42ydqhlnvlvmd4b/share/cmake-3.18/Modules/CMakeDetermineCUDACompiler.cmake:46 (message):
  Could not find compiler set in environment variable CUDAHOSTCXX:
  CC.
Call Stack (most recent call first):
  CMakeLists.txt:264 (enable_language)
```

```console
[lbann-dev-skylake_avx512] vanessen@cgpu12:~/lbann.git/build/gnu.Release.cgpu.nersc.gov/build> printenv CUDAHOSTCXX
CC
```

### Information on your system

<!-- Please include the output of `spack debug report` -->

<!-- If you have any relevant configuration detail (custom `packages.yaml` or `modules.yaml`, etc.) you can add that here as well. -->

CoriGPU

### Additional information

<!-- These boxes can be checked by replacing [ ] with [x] or by clicking them after submitting the issue. -->
- [ ] I have run `spack debug report` and reported the version of Spack/Python/Platform
- [x] I have searched the issues of this repo and believe this is not a duplicate
- [ ] I have run the failing commands in debug mode and reported the output

<!-- We encourage you to try, as much as possible, to reduce your problem to the minimal example that still reproduces the issue. That would help us a lot in fixing it quickly and effectively!

If you want to ask a question about the tool (how to use it, what it can currently do, etc.), try the `#general` channel on our Slack first. We have a welcoming community and chances are you'll get your reply faster and without opening an issue.

Other than that, thanks for taking the time to contribute to Spack! -->

--- a/var/spack/repos/builtin/packages/cuda/package.py
+++ b/var/spack/repos/builtin/packages/cuda/package.py
@@ -86,7 +86,7 @@ class Cuda(Package):
             env.append_path('LD_LIBRARY_PATH', libxml2_home.lib)
 
     def setup_dependent_build_environment(self, env, dependent_spec):
-        env.set('CUDAHOSTCXX', self.compiler.cxx)
+        env.set('CUDAHOSTCXX', dependent_spec.package.compiler.cxx)
 
     def setup_run_environment(self, env):
         env.set('CUDA_HOME', self.prefix)

--- a/var/spack/repos/builtin/packages/cuda/package.py
+++ b/var/spack/repos/builtin/packages/cuda/package.py
@@ -80,15 +80,16 @@ class Cuda(Package):
     depends_on('libxml2', when='@10.1.243:')
 
     def setup_build_environment(self, env):
-        env.set('CUDAHOSTCXX', self.compiler.cxx)
         if self.spec.satisfies('@10.1.243:'):
             libxml2_home  = self.spec['libxml2'].prefix
             env.set('LIBXML2HOME', libxml2_home)
             env.append_path('LD_LIBRARY_PATH', libxml2_home.lib)
 
+    def setup_dependent_build_environment(self, env, dependent_spec):
+        env.set('CUDAHOSTCXX', self.compiler.cxx)
+
     def setup_run_environment(self, env):
         env.set('CUDA_HOME', self.prefix)
-        env.set('CUDAHOSTCXX', self.compiler.cxx)
 
     def install(self, spec, prefix):
         if os.path.exists('/tmp/cuda-installer.log'):

